File: programmers/Level2/문자열 압축.py
#2020 KAKAO BLIND RECRUITMENT
def solution(s):
    zip_length = []
    answer = 0
    if len(s) == 1:  #입력 문자열 s가 1글자 일때 1 반환
        return 1
    
    #1-문자열을 차례대로 분할(한 글자씩 증가)
    for i in range(1, len(s)-1):
        count = 1
        zip_s = ''
        temp_s = s[:i]
            
        #2-반복되는 문자열 검사
        for j in range(i, len(s), i):
            #temp_s가 반복되면 count++
            if s[j:j+i] == temp_s:
                count += 1
            #반복 안될 때
            else:
                #count == 1(반복X)이면 공백
                if count == 1:
                    count = ""
                #압축 결과 생성
                zip_s += (str(count) + temp_s)
                #반복 횟수 초기화
                count = 1
                #비교할 문자열 갱신
                temp_s = s[j:j+i]
        
        #반복되는 문자열 아예 없을 때
        if count == 1:
            count = ""
        zip_s += (str(count) + temp_s)
        #문자열 검사 후 압축 길이 저장
        zip_length.append(len(zip_s))
            
    answer = min(zip_length)
    return answer

File: JAVA_programmers/Level 2/문자열 압축.java
//20210825
//2021 KAKAO BLIND RECRUITMENT


class Solution {
    public int solution(String s) {
        int answer = 0;
        int minLength = Integer.MAX_VALUE;
        StringBuilder sb = new StringBuilder();
        if(s.length() == 1) return 1;
        
        for(int i=1; i<(s.length()/2+1); i++){
            String substr = s.substring(0, i);
            int num = 1, idx = 0;
            
            for(int j=i; j<s.length(); j+=i){    //문자열 비교(j글자씩 반복)
                idx = j+i;
                if(idx > s.length()){
                    substr = s.substring(j-i, s.length());
                    break;
                }
                String temp = s.substring(j, idx);
                if(substr.equals(temp)){
                    num++;
                }else{
                    if(num == 1)
                        sb.append(substr);
                    else
                        sb.append(num+substr);
                    num = 1;
                    idx = j+i;
                    substr = temp;
                }
            }
            if(num == 1)
                sb.append(substr);
            else
                sb.append(num+substr);

            minLength = Math.min(minLength, sb.length());
            sb.setLength(0);
            
        }
        answer = minLength;
        return answer;
    }
}




//20220124
//2021 KAKAO BLIND RECRUITMENT

import java.util.*;


class Solution {
    public int solution(String s) {
        int answer = s.length();
        
        StringBuilder sb = new StringBuilder();
        for(int l=1; l<(s.length()/2)+1; l++){
            int cnt=0;
            String subStr = s.substring(0, l);
            for(int i=0; i<s.length(); i+=l){
                if(i+l > s.length()){
                    subStr = s.substring(i-l, s.length());
                    break;
                }
                
                String temp = s.substring(i, i+l);
                if(subStr.equals(temp)){
                    cnt++;
                }else{
                    if(cnt==1){
                        sb.append(subStr);
                    }else{
                        sb.append(cnt+subStr);
                    }
                    subStr = temp;
                    cnt=1;
                }
            }
            if(cnt==1) sb.append(subStr);
            else sb.append(subStr+cnt);
            answer = Math.min(answer, sb.length());
            sb.setLength(0);
        }
        
        return answer;
    }
}
